### PR TITLE
Wrf hydro test

### PR DESCRIPTION
**Pull request by jmccreight**
Introduce automated testing approach for models/wrf_hydro. 
This is a new feature. The readme describes the changes in this PR
https://github.com/jmccreight/DART/blob/wrf_hydro_test/models/wrf_hydro/hydro_dart_py/hydrodartpy/tests/README.md

### CODE REVIEW

**Comment by hkershaw-brown** [comment]
Can you add a description to this pull request, e.g. 
Does this fix a bug? Add a feature?

**Comment by jmccreight** [comment]
I considered using gitflow conventions... anyway. I updated the description. Sorry, not used to being all formal with DART on github! I will change my ways

**Comment by jmccreight** [comment]
Im noticing that a few files snuck in that are unrelated to the PR, they are what started to motivated it. 
`models/wrf_hydro/experiment_config_files/*yaml`
`models/wrf_hydro/hydro_dart_py/hydrodartpy/core/dartclasses.py`
`models/wrf_hydro/hydro_dart_py/hydrodartpy/core/setup_dart.py`
`models/wrf_hydro/hydro_dart_py/hydrodartpy/core/setup_initial_ens.py`
`models/wrf_hydro/python/perturb/noise_qBtmVertRunoff_additive_fraction.py`

I dont think it's worth the time to back these out, the will be part of my next PR anyway

**Comment by jmccreight** [comment]
Sorry, I also thought I opened this with target jmccreight:master.
My intention was to review it there with Tim and Moha before PR to upstream. Let me know if you want me to move it. Else Im inclined to leave it here.

**Comment by jmccreight** [comment]
I've addressed the dependabot!

File: models/wrf_hydro/hydro_dart_py/setup.py [hunk 1]
```diff
@@ -8,19 +8,7 @@
     url='https://github.com/NCAR/wrf_hydro_dart',
```
Comment (timhoar) [inline models/wrf_hydro/hydro_dart_py/setup.py:8]: James ... this URL is (now) incorrect.   That github repo was made before the entire DART was converted to git.
I believe the correct repo now is the https://github.com/NCAR/DART    (i.e. the one you this PR would merge into).

File: models/wrf_hydro/python/experiment/README.md [hunk 2]
```diff
@@ -57,47 +78,36 @@ echo
...
-
 ### wrf\_hydro\_nwm\_public
 [https://github.com/NCAR/wrf\_hydro\_nwm\_public](https://github.com/NCAR/wrf\_hydro\_nwm\_public)
 The WRF-Hydro model code. The aim is to keep wrf\_hydro\_dart synched with master on this repo.
 
 ### wrf_hydro_py
-[https://github.com/NCAR/wrf\_hydro\_py](https://github.com/NCAR/wrf\_hydro\_py)
-..Python API and tools for managing and running WRF-Hydro and its inputs and outputs (including
-..ensembles. 
+This should boe handled by the install of `hydrodartpy` below. Come here if not or if you need a development install of `wrfhydropy`.
```
Comment (timhoar) [inline models/wrf_hydro/python/experiment/README.md:86]: This should 'be' handled ...

File: models/wrf_hydro/python/experiment/README.md [hunk 3]
```diff
@@ -30,15 +51,12 @@ fi
 # Install and config
 deactivate 2&>1 > /dev/null
 rm -rf $clone_dir
-module load python/3.6.4
-ncar_pylib -c `ncar_pylib -l | tail -n1` $clone_dir
+ncar_pylib -c $py_lib_path $clone_dir
+# The above prints the requsite modules for that env
```
Comment (timhoar) [inline models/wrf_hydro/python/experiment/README.md:55]: 'requisite'  (there's an 'i' in the middle)

**Review by timhoar — CHANGES_REQUESTED** [review]
The stale URL for the ancient private wrf_hydro_dart  repo needs to be changed. As a group we talked about next steps. Stay tuned for more from other group members.

File: models/wrf_hydro/skip_worktree.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
Comment (hkershaw-brown) [inline models/wrf_hydro/skip_worktree.sh:1]: I don't the `skip_worktree.sh` script should be committed.
It seems specific to a local git workflow.

**Comment by hkershaw-brown** [comment]
There are 4 gitignore files for wrf_hydro:

````
./models/wrf_hydro/python/.gitignore
./models/wrf_hydro/hydro_dart_py/hydrodartpy/tests/.gitignore
./models/wrf_hydro/hydro_dart_py/.gitignore
./models/wrf_hydro/.gitignore
````

Can these be consolidated into `./models/wrf_hydro/.gitignore`?

Some of the .gitignore files are missing end of line characters.

File: models/wrf_hydro/hydro_dart_py/hydrodartpy/tests/setup.cfg [hunk 5]
```diff
@@ -0,0 +1,2 @@
+[tool:pytest]
+norecursedirs = data/*
```
Comment (hkershaw-brown) [inline models/wrf_hydro/hydro_dart_py/hydrodartpy/tests/setup.cfg:2]: missing newline

**Comment by mgharamti** [comment]
I tried running the suite on Cheyenne. DART built correctly but the model failed. I believe there is a missing "data" directory in tests. Here is the output: 

gharamti /glade/work/gharamti/DART/DART_development/models/wrf_hydro/hydro_dart_py/hydrodartpy/tests/local/cheyenne 
  [76] $ ./suite_1.sh 
===================================================================================== test session starts =====================================================================================
platform linux -- Python 3.6.8, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- /glade/work/gharamti/python_envs/368/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.6.8', 'Platform': 'Linux-4.12.14-95.51-default-x86_64-with-SuSE-12-x86_64', 'Packages': {'pytest': '4.6.3', 'py': '1.8.0', 'pluggy': '0.12.0'}, 'Plugins': {'datadir-ng': '1.1.0', 'html': '1.21.1', 'metadata': '1.8.0'}, 'JAVA_HOME': '/usr/lib64/jvm/java'}
rootdir: /glade/work/gharamti/DART/DART_development/models/wrf_hydro/hydro_dart_py/hydrodartpy/tests, inifile: setup.cfg
plugins: datadir-ng-1.1.0, html-1.21.1, metadata-1.8.0
collected 3 items                                                                                                                                                                             

test_1_fundamental.py::test_setup_experiment 
expdir=/glade/scratch/gharamti/hydrodartpy_tests/florence_cutout/experiment
rundir=/glade/scratch/gharamti/hydrodartpy_tests/florence_cutout/run
inidir=/glade/scratch/gharamti/hydrodartpy_tests/florence_cutout/initial_ensemble
allobsdir=/glade/scratch/gharamti/hydrodartpy_tests/florence_cutout/obs_seqs
    
Create experiment directory.
DART setup object.
models/wrf_hydro/work: Running "./quickbuild.csh -mpi"
observations/obs_converters/USGS/work: Running "./quickbuild.csh -mpi"
DART successfully compiled into /glade/scratch/gharamti/hydrodartpy_tests/florence_cutout/experiment/dart_build
wrfhydropy.Domain object.
wrfhydropy.Model object.
FAILED
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

config_file = PosixPath('/glade/scratch/gharamti/hydrodartpy_tests/florence_cutout/test.yaml')

    def test_setup_experiment(config_file):
>       setup_return_code = hydrodartpy.setup_experiment(config_file=config_file)

test_1_fundamental.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../core/setup_experiment.py:71: in setup_experiment
    wrf_hydro_sim = setup_wrf_hydro(config)
../core/setup_wrf_hydro.py:37: in setup_wrf_hydro
    compile_options_config_file=config['wrf_hydro']['compile_options_config_file']
/glade/u/home/gharamti/WRF_Hydro/wrf_hydro_py/wrfhydropy/core/model.py:134: in __init__
    self.git_hash = self._get_githash()
/glade/u/home/gharamti/WRF_Hydro/wrf_hydro_py/wrfhydropy/core/model.py:298: in _get_githash
    return get_git_revision_hash(self.source_dir)
/glade/u/home/gharamti/WRF_Hydro/wrf_hydro_py/wrfhydropy/core/model.py:29: in get_git_revision_hash
    cwd=str(the_dir.absolute()))
/glade/u/apps/ch/opt/python/3.6.8/gnu/8.3.0/lib/python3.6/subprocess.py:423: in run
    with Popen(*popenargs, **kwargs) as process:
/glade/u/apps/ch/opt/python/3.6.8/gnu/8.3.0/lib/python3.6/subprocess.py:729: in __init__
    restore_signals, start_new_session)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <subprocess.Popen object at 0x2aab9590c128>, args = ['git', 'branch'], executable = b'git', preexec_fn = None, close_fds = True, pass_fds = ()
cwd = '/glade/work/gharamti/DART/DART_development/models/wrf_hydro/hydro_dart_py/hydrodartpy/tests/data/wrf_hydro_nwm_public/trunk/NDHMS', env = None, startupinfo = None, creationflags = 0
shell = False, p2cread = -1, p2cwrite = -1, c2pread = -1, c2pwrite = 6, errread = -1, errwrite = 6, restore_signals = True, start_new_session = False

    def _execute_child(self, args, executable, preexec_fn, close_fds,
                       pass_fds, cwd, env,
                       startupinfo, creationflags, shell,
                       p2cread, p2cwrite,
                       c2pread, c2pwrite,
                       errread, errwrite,
                       restore_signals, start_new_session):
        """Execute program (POSIX version)"""
    
        if isinstance(args, (str, bytes)):
            args = [args]
        else:
            args = list(args)
    
        if shell:
            args = ["/bin/sh", "-c"] + args
            if executable:
                args[0] = executable
    
        if executable is None:
            executable = args[0]
        orig_executable = executable
    
        # For transferring possible exec failure from child to parent.
        # Data format: "exception name:hex errno:description"
        # Pickle is not used; it is complex and involves memory allocation.
        errpipe_read, errpipe_write = os.pipe()
        # errpipe_write must not be in the standard io 0, 1, or 2 fd range.
        low_fds_to_close = []
        while errpipe_write < 3:
            low_fds_to_close.append(errpipe_write)
            errpipe_write = os.dup(errpipe_write)
        for low_fd in low_fds_to_close:
            os.close(low_fd)
        try:
            try:
                # We must avoid complex work that could involve
                # malloc or free in the child process to avoid
                # potential deadlocks, thus we do all this here.
                # and pass it to fork_exec()
    
                if env is not None:
                    env_list = []
                    for k, v in env.items():
                        k = os.fsencode(k)
                        if b'=' in k:
                            raise ValueError("illegal environment variable name")
                        env_list.append(k + b'=' + os.fsencode(v))
                else:
                    env_list = None  # Use execv instead of execve.
                executable = os.fsencode(executable)
                if os.path.dirname(executable):
                    executable_list = (executable,)
                else:
                    # This matches the behavior of os._execvpe().
                    executable_list = tuple(
                        os.path.join(os.fsencode(dir), executable)
                        for dir in os.get_exec_path(env))
                fds_to_keep = set(pass_fds)
                fds_to_keep.add(errpipe_write)
                self.pid = _posixsubprocess.fork_exec(
                        args, executable_list,
                        close_fds, tuple(sorted(map(int, fds_to_keep))),
                        cwd, env_list,
                        p2cread, p2cwrite, c2pread, c2pwrite,
                        errread, errwrite,
                        errpipe_read, errpipe_write,
                        restore_signals, start_new_session, preexec_fn)
                self._child_created = True
            finally:
                # be sure the FD is closed no matter what
                os.close(errpipe_write)
    
            # self._devnull is not always defined.
            devnull_fd = getattr(self, '_devnull', None)
            if p2cread != -1 and p2cwrite != -1 and p2cread != devnull_fd:
                os.close(p2cread)
            if c2pwrite != -1 and c2pread != -1 and c2pwrite != devnull_fd:
                os.close(c2pwrite)
            if errwrite != -1 and errread != -1 and errwrite != devnull_fd:
                os.close(errwrite)
            if devnull_fd is not None:
                os.close(devnull_fd)
            # Prevent a double close of these fds from __init__ on error.
            self._closed_child_pipe_fds = True
    
            # Wait for exec to fail or succeed; possibly raising an
            # exception (limited in size)
            errpipe_data = bytearray()
            while True:
                part = os.read(errpipe_read, 50000)
                errpipe_data += part
                if not part or len(errpipe_data) > 50000:
                    break
        finally:
            # be sure the FD is closed no matter what
            os.close(errpipe_read)
    
        if errpipe_data:
            try:
                pid, sts = os.waitpid(self.pid, 0)
                if pid == self.pid:
                    self._handle_exitstatus(sts)
                else:
                    self.returncode = sys.maxsize
            except ChildProcessError:
                pass
    
            try:
                exception_name, hex_errno, err_msg = (
                        errpipe_data.split(b':', 2))
                # The encoding here should match the encoding
                # written in by the subprocess implementations
                # like _posixsubprocess
                err_msg = err_msg.decode()
            except ValueError:
                exception_name = b'SubprocessError'
                hex_errno = b'0'
                err_msg = 'Bad exception data from child: {!r}'.format(
                              bytes(errpipe_data))
            child_exception_type = getattr(
                    builtins, exception_name.decode('ascii'),
                    SubprocessError)
            if issubclass(child_exception_type, OSError) and hex_errno:
                errno_num = int(hex_errno, 16)
                child_exec_never_called = (err_msg == "noexec")
                if child_exec_never_called:
                    err_msg = ""
                    # The error must be from chdir(cwd).
                    err_filename = cwd
                else:
                    err_filename = orig_executable
                if errno_num != 0:
                    err_msg = os.strerror(errno_num)
                    if errno_num == errno.ENOENT:
                        err_msg += ': ' + repr(err_filename)
>               raise child_exception_type(errno_num, err_msg, err_filename)
E               FileNotFoundError: [Errno 2] No such file or directory: '/glade/work/gharamti/DART/DART_development/models/wrf_hydro/hydro_dart_py/hydrodartpy/tests/data/wrf_hydro_nwm_public/trunk/NDHMS': '/glade/work/gharamti/DART/DART_development/models/wrf_hydro/hydro_dart_py/hydrodartpy/tests/data/wrf_hydro_nwm_public/trunk/NDHMS'

/glade/u/apps/ch/opt/python/3.6.8/gnu/8.3.0/lib/python3.6/subprocess.py:1364: FileNotFoundError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /glade/u/apps/ch/opt/python/3.6.8/gnu/8.3.0/lib/python3.6/subprocess.py(1364)_execute_child()

**Review by hkershaw-brown — CHANGES_REQUESTED** [review]
added comments about consolidating .gitinore files and text files missing newline character

**Comment by timhoar** [comment]
James agrees that this pull request should be closed and the issues addressed with multiple feature branches and pull requests.

**Comment by timhoar** [comment]
OK James ... lets try again ...

**Comment by jmccreight** [comment]
@mgharamti 
At your leisure, please try this again
```
gharamti /glade/work/gharamti/DART/DART_development/models/wrf_hydro/hydro_dart_py/hydrodartpy/tests/local/cheyenne
[76] $ ./suite_1.sh
```
I can not run this as someone other than myself... so this will be a real help. Running it from scratch (as shown) should take about 400s if it completes successfully. If the compilation is successful, then you can add `--use_existing_build` to the command and get the result in about 100s. This PR was a bit more delayed as I removed use of the scheduler (which requires a person's account key). 

The ultimate success of this may depend strongly on your python (virtual) env. if it all compiles and runs but shows numeric differences in the test_1 output, then that would be a good sign (that the overall machinery is functioning) and we can iterate to solve that. 

LMK the result. happy to chat offline or here
thanks 🙏

**Comment by jmccreight** [comment]
@hkershaw-brown In terms of consolidating the .gitignores, I agree but wanted to keep a spcific .gitignore inside hydro_dart_py.

**Comment by jmccreight** [comment]
@timhoar @hkershaw-brown 
I'm happy to take out models/wrf_hydro/skip_worktree.sh 
do you have any other suggested mechanism for avoiding conflicts in these files that get touched in compilation? e.g. 
```
work/path_names_filter 
work/path_names_model_mod_check
work/path_names_perfect_model_obs
../../assimilation_code/modules/utilities/mpi_utilities_mod.f90
../../assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
```
I know this is something we've discussed in the past.

**Review by mgharamti — APPROVED** [review]
All is good now. The test successfully finished after the changes made and things are in order now.

**Comment by hkershaw-brown** [comment]
> @timhoar @hkershaw-brown
> I'm happy to take out models/wrf_hydro/skip_worktree.sh
> do you have any other suggested mechanism for avoiding conflicts in these files that get touched in compilation? e.g.
> 
> ```
> work/path_names_filter 
> work/path_names_model_mod_check
> work/path_names_perfect_model_obs
> ../../assimilation_code/modules/utilities/mpi_utilities_mod.f90
> ../../assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
> ```
> 
> I know this is something we've discussed in the past.

skip_worktree is a great mechanism, my comment here was on having this specific script in the main DART repo.  It could trip  other users up, who may be working on, for example, perfect_mod_obs .

**Review by hkershaw-brown — CHANGES_REQUESTED** [review]
One of the links to docker hub is broken:

````
WRF-Hydro Domain files in the proper convention 
Currently these can be obtained via
https://hub.docker.com/r/wrfhydro/domains/tags/
````

`https://hub.docker.com/r/wrfhydro/domains/tags/` gives a 404

There are a couple of things that should not make it to the DART master branch, here 
are requested changes: 

The path_name_files files should be reverted to the null_mpi versions. 
skip_worktree.sh script should not be committed.

File: models/wrf_hydro/skip_worktree.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        ../noah/work/path_names_filter
+        ../noah/work/path_names_model_mod_check
+        ../noah/work/path_names_perfect_model_obs )
+
+
+## --skip-worktree and --no-skip-worktree are inverses
+for ff in ${array[@]}; do
+    echo $ff
+    git checkout -- $ff
+    git update-index --no-skip-worktree $ff || exit 1
```
[the remark this answers is not in the document]
Comment (timhoar) [reply models/wrf_hydro/skip_worktree.sh:18]: I just tried  
```
git update-index --assume-unchanged mpi_utilities/mpi_utilities_mod.f90
git update-index --assume-unchanged mpi_utilities/null_mpi_utilities_mod.f90
```

and it seems to have worked beautifully.

File: models/wrf_hydro/work/path_names_filter [hunk 7]
```diff
@@ -22,8 +22,8 @@ assimilation_code/modules/observations/obs_sequence_mod.f90
 assimilation_code/modules/utilities/distributed_state_mod.f90
 assimilation_code/modules/utilities/ensemble_manager_mod.f90
 assimilation_code/modules/utilities/netcdf_utilities_mod.f90
-assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
-assimilation_code/modules/utilities/null_win_mod.f90
+assimilation_code/modules/utilities/mpi_utilities_mod.f90
+assimilation_code/modules/utilities/no_cray_win_mod.f90
```
Comment (timhoar) [inline models/wrf_hydro/work/path_names_filter:26]:
```suggestion
assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
assimilation_code/modules/utilities/null_win_mod.f90
```

File: models/wrf_hydro/work/path_names_model_mod_check [hunk 8]
```diff
@@ -22,8 +22,8 @@ assimilation_code/modules/observations/obs_sequence_mod.f90
 assimilation_code/modules/utilities/distributed_state_mod.f90
 assimilation_code/modules/utilities/ensemble_manager_mod.f90
 assimilation_code/modules/utilities/netcdf_utilities_mod.f90
-assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
-assimilation_code/modules/utilities/null_win_mod.f90
+assimilation_code/modules/utilities/mpi_utilities_mod.f90
+assimilation_code/modules/utilities/no_cray_win_mod.f90
```
Comment (timhoar) [inline models/wrf_hydro/work/path_names_model_mod_check:26]:
```suggestion
assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
assimilation_code/modules/utilities/null_win_mod.f90
```

File: models/wrf_hydro/work/path_names_perfect_model_obs [hunk 9]
```diff
@@ -22,8 +22,8 @@ assimilation_code/modules/observations/obs_sequence_mod.f90
 assimilation_code/modules/utilities/distributed_state_mod.f90
 assimilation_code/modules/utilities/ensemble_manager_mod.f90
 assimilation_code/modules/utilities/netcdf_utilities_mod.f90
-assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
-assimilation_code/modules/utilities/null_win_mod.f90
+assimilation_code/modules/utilities/mpi_utilities_mod.f90
+assimilation_code/modules/utilities/no_cray_win_mod.f90
```
Comment (timhoar) [inline models/wrf_hydro/work/path_names_perfect_model_obs:26]:
```suggestion
assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
assimilation_code/modules/utilities/null_win_mod.f90
```

File: models/wrf_hydro/.gitignore [hunk 10]
```diff
@@ -0,0 +1,7 @@
+../../observations/obs_converters/USGS/work/Makefile
```
Comment (timhoar) [inline models/wrf_hydro/.gitignore:1]:
```suggestion

```

File: models/noah/work/path_names_filter [hunk 11]
```diff
@@ -22,8 +22,8 @@ assimilation_code/modules/observations/obs_sequence_mod.f90
 assimilation_code/modules/utilities/distributed_state_mod.f90
 assimilation_code/modules/utilities/ensemble_manager_mod.f90
 assimilation_code/modules/utilities/netcdf_utilities_mod.f90
-assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
-assimilation_code/modules/utilities/null_win_mod.f90
+assimilation_code/modules/utilities/mpi_utilities_mod.f90
+assimilation_code/modules/utilities/no_cray_win_mod.f90
```
Comment (timhoar) [inline models/noah/work/path_names_filter:26]:
```suggestion
assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
assimilation_code/modules/utilities/null_win_mod.f90
```

File: models/noah/work/path_names_model_mod_check [hunk 12]
```diff
@@ -22,8 +22,8 @@ assimilation_code/modules/observations/obs_sequence_mod.f90
 assimilation_code/modules/utilities/distributed_state_mod.f90
 assimilation_code/modules/utilities/ensemble_manager_mod.f90
 assimilation_code/modules/utilities/netcdf_utilities_mod.f90
-assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
-assimilation_code/modules/utilities/null_win_mod.f90
+assimilation_code/modules/utilities/mpi_utilities_mod.f90
+assimilation_code/modules/utilities/no_cray_win_mod.f90
```
Comment (timhoar) [inline models/noah/work/path_names_model_mod_check:26]:
```suggestion
assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
assimilation_code/modules/utilities/null_win_mod.f90
```

File: models/noah/work/path_names_perfect_model_obs [hunk 13]
```diff
@@ -22,8 +22,8 @@ assimilation_code/modules/observations/obs_sequence_mod.f90
 assimilation_code/modules/utilities/distributed_state_mod.f90
 assimilation_code/modules/utilities/ensemble_manager_mod.f90
 assimilation_code/modules/utilities/netcdf_utilities_mod.f90
-assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
-assimilation_code/modules/utilities/null_win_mod.f90
+assimilation_code/modules/utilities/mpi_utilities_mod.f90
+assimilation_code/modules/utilities/no_cray_win_mod.f90
```
Comment (timhoar) [inline models/noah/work/path_names_perfect_model_obs:26]:
```suggestion
assimilation_code/modules/utilities/null_mpi_utilities_mod.f90
assimilation_code/modules/utilities/null_win_mod.f90
```

File: observations/obs_converters/USGS/.gitignore [hunk 14]
```diff
@@ -0,0 +1 @@
+work/convert_streamflow
```
Comment (timhoar) [inline observations/obs_converters/USGS/.gitignore:1]:
```suggestion
work/convert_streamflow
work/Makefile
```

File: models/wrf_hydro/skip_worktree.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (timhoar) [reply models/wrf_hydro/skip_worktree.sh:18]: So I love it and am going to use it, but agree that the 'skip_worktree.sh' does not belong in the repo.  I am definitely going to mention it in the Wiki and git 'tips and tricks' https://github.com/NCAR/DART/wiki/GIT-Workflow-Tips-&-Tricks

**Review by timhoar — CHANGES_REQUESTED** [review]
Not much help on the python stuff, but here are my comments and suggestions.

**Comment by jmccreight** [comment]
I think all the requested changes are now in place. 
fixing those path_names_* files was a good reminder to run skip_worktree.sh

**Comment by jmccreight** [comment]
Actually a few more changes in the pipeline... this afternoon

**Comment by jmccreight** [comment]
I believe the requested changes are now complete. 
Thanks!

**Pull request merged**